runtime/runtime/src/congestion_control.rs

```diff
@@ -234,24 +234,21 @@ impl ReceiptSinkV2 {
 
         let protocol_version = apply_state.current_protocol_version;
         let shard_layout = epoch_info_provider.shard_layout(&apply_state.epoch_id)?;
-        let shard_ids = if ProtocolFeature::SimpleNightshadeV4.enabled(protocol_version) {
-            shard_layout.shard_ids().collect_vec()
-        } else {
-            self.outgoing_limit.keys().copied().collect_vec()
-        };
-
-        let mut parent_shard_ids = BTreeSet::new();
-        for &shard_id in &shard_ids {
-            let parent_shard_id =
-                shard_layout.try_get_parent_shard_id(shard_id).map_err(Into::<EpochError>::into)?;
-            let Some(parent_shard_id) = parent_shard_id else {
-                continue;
+        let (shard_ids, parent_shard_ids) =
+            if ProtocolFeature::SimpleNightshadeV4.enabled(protocol_version) {
+                (
+                    shard_layout.shard_ids().collect_vec(),
+                    shard_layout.get_parent_shard_ids().map_err(Into::<EpochError>::into)?,
+                )
+            } else {
+                (self.outgoing_limit.keys().copied().collect_vec(), BTreeSet::new())
             };
-            if parent_shard_id == shard_id {
-                continue;
-            }
-            parent_shard_ids.insert(parent_shard_id);
-        }
+
+        // There mustn't be any shard ids in both the parents and the current
+        // shard ids. If this happens the same buffer will be processed twice.
+        debug_assert!(
+            parent_shard_ids.intersection(&shard_ids.clone().into_iter().collect()).count() == 0
+        );
 
```

Comment:
Would it be possible to ban using `state_update.trie`? Make it private or something. Or at least rename it to `state_update.immutable_pre_state_trie`.
 I don't see any real reason to ever read the pre-state trie instead of the current state. It's a footgun waiting to go off. Forwarding from the same buffer wouldn't be a problem if the iterator used `state_update` instead of `state_update.trie`. I guess it changes trie reads so it would be a protocol upgrade, but it shouldn't be too bad.
 
 I'd prefer that over delicate logic that needs shard ids to never repeat.

Reply:
Can you file an issue an tag me on it please? I don't think I'll have the time to work on this in the near future but overall sounds good. I'm curious what was the original reason for using the pre-state trie.